test/e2e/rule_test.go

```diff
@@ -158,19 +160,26 @@ func reloadRulesSignal(t *testing.T, r e2e.InstrumentedRunnable) {
 }
 
 func checkReloadSuccessful(t *testing.T, ctx context.Context, endpoint string, expectedRulegroupCount int) {
-	req, err := http.NewRequestWithContext(ctx, "GET", "http://"+endpoint+"/api/v1/rules", ioutil.NopCloser(bytes.NewReader(nil)))
-	testutil.Ok(t, err)
-	resp, err := http.DefaultClient.Do(req)
-	testutil.Ok(t, err)
-	testutil.Equals(t, 200, resp.StatusCode)
+	var data = rulesResp{}
+	testutil.Ok(t, runutil.Retry(5*time.Second, ctx.Done(), func() error {
+		req, err := http.NewRequestWithContext(ctx, "GET", "http://"+endpoint+"/api/v1/rules", ioutil.NopCloser(bytes.NewReader(nil)))
+		testutil.Ok(t, err)
+		resp, err := http.DefaultClient.Do(req)
+		testutil.Ok(t, err)
+		testutil.Equals(t, 200, resp.StatusCode)
 
-	body, _ := ioutil.ReadAll(resp.Body)
-	testutil.Ok(t, resp.Body.Close())
+		body, _ := ioutil.ReadAll(resp.Body)
+		testutil.Ok(t, resp.Body.Close())
 
-	var data = rulesResp{}
+		testutil.Ok(t, json.Unmarshal(body, &data))
+		testutil.Equals(t, "success", data.Status)
 
-	testutil.Ok(t, json.Unmarshal(body, &data))
-	testutil.Equals(t, "success", data.Status)
+		// Retry until count matches.
+		if len(data.Data.Groups) == expectedRulegroupCount {
+			return nil
+		}
+		return errors.New("different number of rulegroups")
```

Comment:
Maybe we can add the number of expected rulegroups vs. number of retrieved rulegroups? Something like:
```suggestion
		return errors.Newf("different number of rulegroups: expected %d, got %d", expectedRulegroupCount, len(data.Data.Groups))
```